c10/core/TensorImpl.cpp

```diff
@@ -184,6 +177,44 @@ TensorImpl::TensorImpl(
   // Caffe2 operators create Storages with default devices.
 }
 
+TensorImpl::TensorImpl(
+    const TensorImpl& that,
+    intrusive_ptr<impl::cow::Simulator> copy_on_write_simulator)
+    : TensorImpl(
+          Storage(that.storage()),
+          that.key_set(),
+          that.dtype(),
+          std::move(copy_on_write_simulator)) {}
+
+// NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+TensorImpl::TensorImpl(
+    Storage&& storage,
+    DispatchKeySet key_set,
+    const caffe2::TypeMeta data_type,
+    intrusive_ptr<impl::cow::Simulator> copy_on_write_simulator)
+    : intrusive_ptr_target(),
+      storage_(std::move(storage)),
+      numel_(0),
+      data_type_(data_type),
+      device_opt_(storage_.device()),
+      key_set_(key_set - c10::python_ks), // See [Note: Python key removal]
+      copy_on_write_simulator_(std::move(copy_on_write_simulator)) {
+  init_bitfields();
+  // Inference tensor doesn't have version counter.
+  if (!is_inference()) {
+    version_counter_ = VariableVersion(/*version=*/0);
+  }
+}
+
+intrusive_ptr<TensorImpl> TensorImpl::take_view() const {
```

Comment:
i haven't done this yet. do you recommend expanding my unit test with all the view ops to make sure they are covered?

Reply:
We should have OpInfos for view ops, you just have to make sure you send them CoW'd tensors in a custom test.

Reply:
i've migrated the calls of ImplType::VIEW to use take_view. Right now i'm not supporting tensor subclasses, so those calls still exist.

Reply:
As I mentioned in another comment, I think you need to handle tensor subclasses. And because you need to handle tensor subclasses, I think the `take_view` method as done here is not good design. You want to set everything up in the constructors. What would have been more logical to me would have been to modify the VIEW constructor to accept shadow storage, and then update all the call sites to do the right thing.